### ISSUE TITLE
Recompress and Store decoded ASTC

### ISSUE THREAD

**Issue by AsrielDewitt** [yuzu-emu/yuzu#11691]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What feature are you suggesting?

Storing the decoded ASTC textures and recompressed as BC7 for specific games

### Why would this feature be useful?

Speed up texture load times by not decode ASTC textures every time and just load already compressed BC7

**Comment by Squall-Leonhart** [comment]
i've been told this is near impossible and likely requires per-game patches to pull off anyway.

**Comment by boludoz** [comment]
> i've been told this is near impossible and likely requires per-game patches to pull off anyway.

Nothing is impossible with code, you just have to invent solutions. I don't understand much, but maybe a hash for each texture?

**Comment by FernandoS27** [comment]
Is not impossible but due to size, there may be a need of a hybrid cache. 

I do wanted to do this at some point as it will be needed for texture replacement.

**Comment by Squall-Leonhart** [comment]
> > i've been told this is near impossible and likely requires per-game patches to pull off anyway.
> 
> Nothing is impossible with code, android emulators do it, you just have to invent solutions. I don't understand much, but maybe a hash for each texture? In any case, we would have to evaluate if it would really work.

i did consider an adaption of the GlideN64 streaming disk cache mechanism.

> We could allow you to allocate an LRU of RAM.

This was a source of stutters and massively blown out loading times with the older Rice/Glide64 method of loading texture replacements.

**Comment by Adamillo** [comment]
What if we made an ASTC cache instead where yuzu stores a cache for the converted ASTC textures?

**Comment by FernandoS27** [comment]
> What if we made an ASTC cache instead where yuzu stores a cache for the converted ASTC textures?

Yeah the issue, though is that IO operations have delays which can add stutter, especially in hard drives.

I've been thinking in this idea for a while. I know that we could try using the CASC system from Blizzard Entertainment for IO and adapt it to direct GPU fetch.

A big issue, though is that textures coming from the game are swizzled and the alignment padding can contain garbage, messing with the hash calculator. So they must be at least unswizzled before being hashed to check the cache.

**Comment by boludoz** [comment]
> > What if we made an ASTC cache instead where yuzu stores a cache for the converted ASTC textures?
> 
> Yeah the issue, though is that IO operations have delays which can add stutter, especially in hard drives. This can always be delegated to worker threads but it will still add stutter.
> 
> I've been thinking in this idea for a while. I know that we could try using the CASC system from Blizzard Entertainment for IO and adapt it to direct GPU fetch.
> 
> A big issue, though is that textures coming from the game are swizzled and the alignment padding can contain garbage, messing with the hash calculator. So they must be at least unswizzled before being hashed to check the cache.

Isn't it possible to refer to the address within the game? Instead of hashing the texture, which would be expensive. That never changes, I speak from ignorance.

**Comment by brujo5** [comment]
Sorry for the interruption but, how is Skyline doing then?  game loading, texture loading, etc. is instantaneous, it takes almost nothing in that emulator compared to yuzu and eggy and the RAM consumed is minimal that even phones with 6GB of RAM have no problem

**Comment by boludoz** [comment]
> Sorry for the interruption but, how is Skyline doing then? game loading, texture loading, etc. is instantaneous, it takes almost nothing in that emulator compared to yuzu and eggy and the RAM consumed is minimal that even phones with 6GB of RAM have no problem

That's because skyline is specially optimized for ARM, plus the Nintendo Switch is also ARM, and has a special feature called NCE. But I honestly think it has nothing to do with this.

**Comment by boludoz** [comment]
> Contributor

I imagine that the complexity here must be much greater, sir.

**Comment by boludoz** [comment]
> > > > What if we made an ASTC cache instead where yuzu stores a cache for the converted ASTC textures?
> > > 
> > > 
> > > Yeah the issue, though is that IO operations have delays which can add stutter, especially in hard drives. This can always be delegated to worker threads but it will still add stutter.
> > > I've been thinking in this idea for a while. I know that we could try using the CASC system from Blizzard Entertainment for IO and adapt it to direct GPU fetch.
> > > A big issue, though is that textures coming from the game are swizzled and the alignment padding can contain garbage, messing with the hash calculator. So they must be at least unswizzled before being hashed to check the cache.
> > 
> > 
> > Isn't it possible to refer to the address within the game? Instead of hashing the texture, which would be expensive. That never changes, I speak from ignorance. I think pcsx2 does something similar to replace textures and saves them with a hashed name.
> 
> Old school emulators can do that because they are deterministic. Because the switch has a multi core modern CPU, it's really hard to make the emulation deterministic without losing performance. Thus, the games in yuzu always have different addresses.

I imagine that the complexity here must be much greater, sir.

**Comment by AsrielDewitt** [comment]
> > > > > What if we made an ASTC cache instead where yuzu stores a cache for the converted ASTC textures?
> > > > 
> > > > 
> > > > Yeah the issue, though is that IO operations have delays which can add stutter, especially in hard drives. This can always be delegated to worker threads but it will still add stutter.
> > > > I've been thinking in this idea for a while. I know that we could try using the CASC system from Blizzard Entertainment for IO and adapt it to direct GPU fetch.
> > > > A big issue, though is that textures coming from the game are swizzled and the alignment padding can contain garbage, messing with the hash calculator. So they must be at least unswizzled before being hashed to check the cache.
> > > 
> > > 
> > > Isn't it possible to refer to the address within the game? Instead of hashing the texture, which would be expensive. That never changes, I speak from ignorance. I think pcsx2 does something similar to replace textures and saves them with a hashed name.
> > 
> > 
> > Old school emulators can do that because they are deterministic. Because the switch has a multi core modern CPU, it's really hard to make the emulation deterministic without losing performance. Thus, the games in yuzu always have different addresses.
> 
> I imagine that the complexity here must be much greater, sir. From my ignorance I thought that you could see the file system addresses for each texture.
> 
> Perhaps the gain is minimal then because of the blockages caused by accessing the same element at the same time, in addition to having to access the hard drive.

damm, if only there's a way to access the textures directly without IO delay, like... Direct Storage..huh. Too bad no Vulkan equivelent

**Comment by Squall-Leonhart** [comment]
> I imagine that the complexity here must be much greater, sir. From my ignorance I thought that you could see the file system addresses for each texture.

A tool that extracts, converts to BC# (heuristics tested) and then builds a mod might better here.

**Comment by Adamillo** [comment]
> > i've been told this is near impossible and likely requires per-game patches to pull off anyway.
> 
> Nothing is impossible with code, android emulators do it, you just have to invent solutions. I don't understand much, but maybe a hash for each texture? In any case, we would have to evaluate if it would really work.

The reason that Switch Emulators on Android can load ASTC textures instantly is because of the fact that most Android GPU's natively support ASTC textures, so they don't have to deal with converting any textures. Same goes for the Intel iGPU's on PC's, they also natively support ASTC

**Comment by Squall-Leonhart** [comment]
unfortuantely UHD 700 and Arc Xe drop support for ASTC.

**Comment by AsrielDewitt** [comment]
> > I imagine that the complexity here must be much greater, sir. From my ignorance I thought that you could see the file system addresses for each texture.
> 
> A tool that extracts, converts to BC# (heuristics tested) and then builds a mod might better here.

Might Need an initially load though, and I don't think all games even stored textures the same way
